src/vmhistogram.rs

```diff
@@ -0,0 +1,328 @@
+//! Rust reimplementation of https://github.com/VictoriaMetrics/metrics/blob/ae1e9d8058de94e5622ccee515ece7992d67d209/histogram.go
+// Originally distributed under the following license:
+// The MIT License (MIT)
+//
+// Copyright (c) 2019 VictoriaMetrics
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+use std::{
+    mem::MaybeUninit,
+    sync::{Arc, LazyLock},
+};
+
+use parking_lot::Mutex;
+
+use crate::{
+    core::{Collector, Desc, Describer, Metric, MetricVec, MetricVecBuilder},
+    proto::{self, LabelPair},
+    value::make_label_pairs,
+    Opts,
+};
+
+const E_10_MIN: i32 = -9;
+const E_10_MAX: i32 = 18;
+const BUCKETS_PER_DECIMAL: usize = 18;
+const DECIMAL_BUCKETS_COUNT: usize = (E_10_MAX - E_10_MIN) as usize;
+const BUCKETS_COUNT: usize = DECIMAL_BUCKETS_COUNT * BUCKETS_PER_DECIMAL;
+static BUCKET_MULTIPLIER: LazyLock<f64> =
```

Comment:
interesting that `powf` isn't const (they perhaps don't want to use `libm` at compile time?) -- can you leave a comment?

Reply:
yeah, rust doesn't actually support doing anything float related at compile time yet: https://users.rust-lang.org/t/why-is-f64-powf-not-a-const-fn/31093/5